File: java/affine-cipher/src/main/java/AffineCipher.java
import java.util.function.UnaryOperator;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class AffineCipher {
    private static final String ALPHABET = "abcdefghijklmnopqrstuvwxyz";
    private static final int ENCODED_GROUP_SIZE = 5;

    String encode(String text, int a, int b) {
        validateKey(a);

        String encoded = __code(text, c -> {
                int x = ALPHABET.indexOf(c);
                if (x != -1)
                    c = ALPHABET.charAt(
                            Math.floorMod(
                                    a * x + b,
                                    ALPHABET.length()
                            )
                    );
                return c;
        });
        return groupsOf(ENCODED_GROUP_SIZE, encoded);
    }

    String decode(String text, int a, int b) {
        validateKey(a);
        int a_inv = mmi(a, ALPHABET.length());

        return __code(text, c -> {
                int y = ALPHABET.indexOf(c);
                if (y != -1)
                    c = ALPHABET.charAt(
                            Math.floorMod(
                                    a_inv * (y - b),
                                    ALPHABET.length()
                            )
                    );
                return c;
        });
    }

    private void validateKey(int a) {
        if (gcd(a, ALPHABET.length()) != 1)
            throw new IllegalArgumentException("Error: keyA and alphabet size must be coprime.");
    }

    private String __code(String text, UnaryOperator<Character> op) {
        return text
                .replaceAll("[^\\p{Alnum}]", "")
                .toLowerCase()
                .chars()
                .mapToObj(c -> op.apply((char) c))
                .collect(StringBuilder::new, StringBuilder::append, StringBuilder::append)
                .toString();
    }

    private String groupsOf(int n, String text) {
        Stream.Builder<String> sb = Stream.builder();
        Matcher m = Pattern.compile(".{1," + n + "}").matcher(text);
        while (m.find())
            sb.accept(m.group());
        return sb.build().collect(Collectors.joining(" "));
    }

    private int mmi(int a, int m) {
        // not too efficient
        int n = 0;
        while (Math.floorMod(a * n, m) != 1)
            n++;
        return n;
    }

    private int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }
}

File: kotlin/affine-cipher/src/main/kotlin/AffineCipher.kt
object AffineCipher {
    private val Alphabet = ('a'..'z').joinToString("")

    fun encode(input: String, a: Int, b: Int): String {
        require(a.isCoprimeWith(Alphabet.length)) { "a and m must be coprime." }
        return code(input) { x -> a * x + b }
                .chunked(5)
                .joinToString(" ")
    }

    fun decode(input: String, a: Int, b: Int): String {
        require(a.isCoprimeWith(Alphabet.length)) { "a and m must be coprime." }
        val aInv = a.mmi(Alphabet.length)
        return code(input) { y -> aInv * (y - b) }
    }

    private fun code(text: String, transform: (Int) -> Int): String =
        text.filter(Char::isLetterOrDigit)
            .map(Char::toLowerCase)
            .map { c ->
                when (val i = Alphabet.indexOf(c)) {
                    -1 -> c
                    else -> Alphabet[transform(i).floorMod(Alphabet.length)]
                }
            }
            .joinToString("")
}

fun Int.isCoprimeWith(other: Int): Boolean = this.gcd(other) == 1

// could use java.lang.BigInteger::gcd
tailrec fun Int.gcd(b: Int): Int = if (b == 0) this else b.gcd(this % b)

// return the modulus in the range (0 until this)
// could use java.lang.Math.floorMod
fun Int.floorMod(m: Int): Int = (this % m + m) % m

// could use java.lang.BigInteger::modInverse
fun Int.mmi(m: Int): Int = (0 until m).first { (this * it).floorMod(m) == 1 }
